### ISSUE TITLE
Describe how to link in a web project w/o bundling

### ISSUE THREAD

**Issue by mighdoll** [wgsl-tooling-wg/wesl-spec#22]
Some js projects want to be able to build w/minimal tooling - essentially running nodejs and javascript/typescript only - no importing of metadata files, no bundler, no additional tools to run during the build. 

Describe how to link in a low tooling world, and what if anything won't work as well in this environment (e.g. language server limitations, missing package versioning metadata)

Here's a perspective for example from vis.gl:
> I think it is fine if your project support both files and strings and you leave the choice to users.
> *  I would not be using separate files in my projects,
> *  and I would expect to run into overwhelming bundling issues if I tried to use published libraries that include shaders in separate files (unless they go through a build step before publishing that inlines the shaders into strings).

vis.gl mentioned that they dropped our spiritual predecessor glsify due to its hard dependency on tooling and the ensuing maintenance problems that caused.

**Comment by ncthbrt** [comment]
Still think a URL driven model could work for loads/imports on web

**Comment by mighdoll** [comment]
I imagine you bring up URLs because if every import referenced a public URL, a runtime linker library could load packages without other tools. But in a world where wesl packages are in npm, how does also having URL loading help?

**Comment by ncthbrt** [comment]
My main thought was that a runtime linker could use the urls to load from a
cdn or the backend or whatever without additional machinery or tooling.
Whereas tying it to a package manager makes this difficult

**Comment by mighdoll** [comment]
> My main thought was that a runtime linker could use the urls to load from a cdn or the backend or whatever without additional machinery or tooling. Whereas tying it to a package manager makes this difficult

Yep it adds difficulty, but I think web folks will expect to use npm.

**Comment by k2d222** [comment]
> Still think a URL driven model could work for loads/imports on web

I have mixed opinions about this -- look at [Vite](https://vitejs.dev/) for example.

On one hand it supports importing all kinds of files, like CSS or JSON, even images which is both very cool and very useful,
But on the other hand the source files become tied to Vite and incompatible with other bundlers.

[my first import proposal](https://docs.google.com/document/d/11dngLll2foS0TA5NZGP9lb8IC4ZmK_FVKhSSU8Qjn14/edit?usp=sharing) was going even further, where basically all import path resolutions are implementation-defined. This allows the same kind of super cool stuff that Vite does (and more, see some examples on the proposal)

@mighdoll and @stefnotch made me realize that whatever is not unspecified by the spec, subjects the ecosystem to fragmentation as implementations diverge and grow more complex. Everyone hates the fragmentation of the web ecosystem for example.

So... At this point I am not sure. Maybe standardizing two syntaxes: one with double quotes and where resolution is left unspecified by the spec, and one with normal paths and separators, well-defined by the spec. A published library would not be allowed to use the quoted imports. I think this is a good compromise.

**Comment by mighdoll** [comment]
opened #23 to discuss urls in import statements.

**Comment by s-ol** [comment]
I also have a use-case for package-manager-less runtime package loading:

I've built a "shader host" for livecoding. It's essentially a server that you send shader code to and get rendered video out. The tool is written in Rust so it could load WESL libraries via Cargo, but that of course requires the dependencies to be known at build time, whereas my built binary should be completely content-agnostic.
It's at runtime that the user should be able to launch my binary and indicate a way to load *their* package delendencies (perhaps indicating a library directory, that *could* itself be managed using e.g. npm) in the context of their project.

Also note that since this is livecoding, the user code never goes through a "build phase" to which the bundling could be relegated.

For now I've built some custom resolvers for the wesl crate that handle this by rewriting the import paths, which is *fine* but could be helped.

I do see the benefit in *not* supporting "raw" loading from the filesystem as a feature, the current state provides a good nudge towards the package managers, which helps stabilize the ecosystem with strong packaging guidelines. When raw access becomes the option with least friction, it might become the go-go option despite drawbacks (no build system integration / caching) and fragment the bundling practices.

**Comment by mighdoll** [comment]
> I also have a use-case for package-manager-less runtime package loading:
> 
> Also note that since this is livecoding, the user code never goes through a "build phase" to which the bundling could be relegated.
>
> I've built a "shader host" for livecoding. 

on the -js side, wgsl-play/wgsl-edit currently support live loading, see `fetch-libs` and `fetch-sources` in the [README](https://github.com/webgpu-tools/wesl-js/blob/main/packages/wgsl-play/README.md#attributes). You can type in a package name and it'll fetch from npm, you can load files from static hosting, etc. I was modeling the -js resolver path on the -rs linker originally, so I wonder if something could be added or if you could build it on the outside for wesl-rs. It'd be wild to have an npm packages load live into wesl-rs!

**Comment by s-ol** [comment]
>  I wonder if something could be added or if you could build it on the outside for wesl-rs

You definitely can, it just feels a little clumsy having to meddle with the paths like this when there is a `FileResolver` that handles only `package::` paths:

```rs
pub struct ShaderSources {
    virtual_resolver: VirtualResolver<'static>,
    file_resolver: Option<FileResolver>,
}

impl Resolver for ShaderSources {
    fn resolve_source<'a>(&'a self, path: &ModulePath) -> Result<Cow<'a, str>, ResolveError> {
        match path.origin {
            PathOrigin::Absolute => self.virtual_resolver.resolve_source(path),
            PathOrigin::Package(ref pname) => {
                if let Some(ref files) = self.file_resolver {
                    let fpath = ModulePath {
                        origin: PathOrigin::Absolute,
                        components: iter::once(pname)
                            .chain(path.components.iter())
                            .cloned()
                            .collect(),
                    };
                    files.resolve_source(&fpath)
                } else {
                    Err(ResolveError::ModuleNotFound(
                        path.clone(),
                        "no library directory found".into(),
                    ))
                }
            }
            _ => Err(ResolveError::ModuleNotFound(
                path.clone(),
                "can't resolve relative module".into(),
            )),
        }
    }
}
```
